### PR TITLE
1/2 - Commonwell Document Query

### PR DESCRIPTION
**Pull request by leite08**
Ref: https://github.com/metriport/metriport-internal/issues/227

### Dependencies

- Upstream: none
- Downstream: https://github.com/metriport/metriport/pull/91

### Description

WIP of Commonwell Document Consumption, includes only the first part: doc query

Also fixes other issues with compilation, lint, prettier.

A couple of adjustments on existing models.

### Release Plan

- ⚠️ "Squash & merge" as we have some temporary commits on the history

### CODE REVIEW

File: packages/packages/commonwell-cert-runner/src/document-consumption.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import {
+  getPatientStrongIds,
+  getPersonIdFromSearchByPatientDemo,
+} from "@metriport/commonwell-sdk/lib/common/util";
+import { docPatient, docPerson, metriportSystem } from "./payloads";
+
+// Document Consumption
+// https://commonwellalliance.sharepoint.com/sites/ServiceAdopter/SitePages/Document-Consumption-(SOAP,-REST).aspx
+
+// TODO Remove logs with the prefix '...'
```
Comment (leite08) [inline packages/packages/commonwell-cert-runner/src/document-consumption.ts:19]: Will clean these up on the upcoming PR

File: packages/packages/commonwell-cert-runner/src/link-management.ts [hunk 2]
```diff
@@ -7,8 +7,8 @@ import { patient, identifier, personStrongId } from "./payloads";
 // https://commonwellalliance.sharepoint.com/sites/ServiceAdopter/SitePages/Patient-Management-(PIX,-REST).aspx
 
 export async function linkManagement(commonWell: CommonWell, queryMeta: RequestMetadata) {
-  // C5: ​Levels of Link Assurance
-  console.log(`>>> ​C5a : Link a Patient to a Person upgrading from LOLA 1 to LOLA 2.`);
+  // C5: Levels of Link Assurance
+  console.log(`>>> C5a : Link a Patient to a Person upgrading from LOLA 1 to LOLA 2.`);
```
Comment (leite08) [inline packages/packages/commonwell-cert-runner/src/link-management.ts:11]: These changes where Github doesn't indicate exactly what changed are related to invalid whitespaces (lint rule `no-irregular-whitespace`).

File: packages/packages/commonwell-cert-runner/src/org-management.ts [hunk 3]
```diff
@@ -42,26 +42,23 @@ export async function orgManagement(commonWell: CommonWell, queryMeta: RequestMe
   console.log(respReplaceCertificateForOrg);
 
   console.log(`>>> Get certificate from org`);
-  const respGetCertificateFromOrg = await commonWell.getCertificateFromOrg(queryMeta, orgId);
-  console.log(respGetCertificateFromOrg);
+  const respGetCertificatesFromOrg = await commonWell.getCertificatesFromOrg(queryMeta, orgId);
```
Comment (leite08) [inline packages/packages/commonwell-cert-runner/src/org-management.ts:45]: Compilation errors from `develop`

File: packages/packages/commonwell-sdk/src/models/document.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+});
+
+const statusSchema = z.enum(["current", "superceded", "entered in error"]);
+
+// Main Clinical Acts Documented
+const eventSchema = z.object({
+  text: z.string(), // CodeableConcept
+});
+
+export const contentSchema = z.object({
```
Comment (Goncharo) [inline packages/packages/commonwell-sdk/src/models/document.ts:37]: It looks like we're missing a lot of fields in this model? Why is that?

Also let's keep the names consistent with CW, perhaps `documentReferenceSchema` ?

**Review by Goncharo — CHANGES_REQUESTED** [review]

File: packages/packages/commonwell-cert-runner/src/document-consumption.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+export async function documentConsumption(commonWell: CommonWell, queryMeta: RequestMetadata) {
+  // E1: Document Query
+  console.log(`>>> E1c: Query for documents using FHIR (REST)`);
+
+  console.log(`... Search for Patient...`, {
```
Comment (Orta21) [inline packages/packages/commonwell-cert-runner/src/document-consumption.ts:25]: Why the console log for search for patient?

File: packages/packages/commonwell-sdk/src/client/commonwell.ts [hunk 6]
```diff
@@ -620,6 +645,22 @@ export class CommonWell {
     return;
   }
 
+  /**
+   * Queries a Document
```
[the remark this answers is not in the document]
Comment (leite08) [reply packages/packages/commonwell-sdk/src/client/commonwell.ts:649]: Ended up updating this on the downstream PR: https://github.com/metriport/metriport/pull/91/files#diff-7c4b22db85224ac3dd9dc8d4bf23bd7c9db78e160b975001b290c9720d729da7R652

File: packages/packages/commonwell-cert-runner/src/document-consumption.ts (hunk 5 above)
Comment (leite08) [reply packages/packages/commonwell-cert-runner/src/document-consumption.ts:25]: To be removed on the upcoming PR as indicated [here](https://github.com/metriport/metriport/pull/89/files/92021604b3ea944c165ca3283380de48f437672d#r1091161125). Kept to help debugging while I work on the next one.

File: packages/packages/commonwell-sdk/src/client/commonwell.ts [hunk 7]
```diff
@@ -620,6 +645,22 @@ export class CommonWell {
...
+   * See: https://specification.commonwellalliance.org/services/data-broker/cha-broker-api-reference#104-document-query
+   *
+   * @param meta       Metadata about the request.
+   * @param subjectId  The ID as defined by the specification: [system]|[code] where 'system' is the OID
```
[the remark this answers is not in the document]
Comment (leite08) [reply packages/packages/commonwell-sdk/src/client/commonwell.ts:653]: Yessss, good catch!

File: packages/packages/commonwell-sdk/src/common/util.ts [hunk 8]
```diff
@@ -1,14 +1,48 @@
...
+export function getPersonIdFromSearchByPatientDemo(object: PersonSearchResp): string | undefined {
+  if (!object._embedded || !object._embedded.person) return undefined;
+  const embeddedPersons = object._embedded.person.filter(p => p.enrolled);
+  if (embeddedPersons.length < 1) return undefined;
+  if (embeddedPersons.length > 1) {
+    console.log(`Found more than one person, using the first one: `, object);
+  }
+  const person = embeddedPersons[0];
+  const url = person._links?.self?.href;
+  return getPersonIdFromUrl(url);
```
[the remark this answers is not in the document]
Comment (leite08) [reply packages/packages/commonwell-sdk/src/common/util.ts:25]: Good point! Will do!

File: packages/packages/commonwell-sdk/src/models/document.ts (hunk 4 above)
Comment (leite08) [reply packages/packages/commonwell-sdk/src/models/document.ts:37]: I'm reviewing these... I'm using the actual return from the endpoint as reference and found some discrepancies, so I didn't want to constrain too much, but I think its fine. Reviewing...

Comment (leite08) [reply packages/packages/commonwell-sdk/src/models/document.ts:49]: Reviewed one by one. There's a fair gap between the spec and what we get as a response when querying documents - I'm even thinking that spec is for a document reference, and not the document itself?

This is an example of an actual response:

```json
{
    "resourceType": "Bundle",
    "entry": [
        {
            "id": "urn:uuid:e7f03f5a-db17-43c2-9ad7-d6f941bf64e0",
            "content": {
                "resourceType": "DocumentReference",
                "contained": [
                    {
                        "resourceType": "Organization",
                        "id": "orgRef1",
                        "name": "Metriport-OrgB-1651"
                    },
                    {
                        "resourceType": "Practitioner",
                        "id": "authRef1",
                        "organization": {
                            "reference": "#orgRef1"
                        }
                    }
                ],
                "masterIdentifier": {
                    "system": "urn:ietf:rfc:3986",
                    "value": "2.16.840.1.113883.3.3330.8889429.1651.2^89765..1"
                },
                "identifier": [
                    {
                        "use": "official",
                        "system": "urn:ietf:rfc:3986",
                        "value": "urn:uuid:e7f03f5a-db17-43c2-9ad7-d6f941bf64e0"
                    }
                ],
                "subject": {
                    "reference": "#patRef1"
                },
                "type": {
                    "coding": [
                        {
                            "system": "http://loinc.org/",
                            "code": "34133-9",
                            "display": "Summarization of episode note"
                        }
                    ]
                },
                "author": [
                    {
                        "reference": "#authRef1"
                    }
                ],
                "indexed": "2023-01-31T16:56:50+00:00",
                "status": "current",
                "description": "Metriport-OrgB-1651 Summarization Of Episode Note",
                "mimeType": "application/octet-stream",
                "location": "https://integration.rest.api.commonwellalliance.org/v2/Binary/H4sIAAAAAAAEAJXLOw6AIBBF0d1YvjigOBTGwsoF2NoIiRQwfqBw97oFcsubMx0S_SwxlhTyu7hRgQy4a0Eg0swaf1q3YGbbKfvvnqCa25_yhCz3u6ZwFV9Bnewl-pRr4cZ2MD1AHwauaVW1AAAA0",
                "context": {
                    "event": [
                        {
                            "text": "Problem"
                        },
                        {
                            "text": "drug allergy"
                        },
                        {
                            "text": "Nausea"
                        },
                        {
                            "text": "Depressive disorder"
                        },
                        {
                            "text": "Hypothyroidism"
                        },
                        {
                            "text": "Sinusitis"
                        },
                        {
                            "text": "CT of paranasal sinuses"
                        },
                        {
                            "text": "Left maxillary sinus mass"
                        },
                        {
                            "text": "never a smoker"
                        },
                        {
                            "text": "Vital signs"
                        },
                        {
                            "text": "Neoplasm of accessory sinus"
                        },
                        {
                            "text": "Finding of functional performance and activity"
                        },
                        {
                            "text": "Cane"
                        },
                        {
                            "text": "Cognitive function finding"
                        },
                        {
                            "text": "Short term memory deficit"
                        }
                    ],
                    "period": {
                        "start": "2012-10-02T04:00:00Z"
                    }
                }
            }
        },
        {
            "id": "urn:uuid:b8fa0f9a-9915-40ce-b0de-478d22575cc3",
            "content": {
                "resourceType": "DocumentReference",
                "contained": [
                    {
                        "resourceType": "Organization",
                        "id": "orgRef2",
                        "name": "Metriport-OrgA-1651"
                    },
                    {
                        "resourceType": "Practitioner",
                        "id": "authRef2",
                        "organization": {
                            "reference": "#orgRef2"
                        }
                    }
                ],
                "masterIdentifier": {
                    "system": "urn:ietf:rfc:3986",
                    "value": "2.16.840.1.113883.3.3330.8889429.1651.1^89765..1"
                },
                "identifier": [
                    {
                        "use": "official",
                        "system": "urn:ietf:rfc:3986",
                        "value": "urn:uuid:b8fa0f9a-9915-40ce-b0de-478d22575cc3"
                    }
                ],
                "subject": {
                    "reference": "#patRef2"
                },
                "type": {
                    "coding": [
                        {
                            "system": "http://loinc.org/",
                            "code": "34133-9",
                            "display": "Summarization of episode note"
                        }
                    ]
                },
                "author": [
                    {
                        "reference": "#authRef2"
                    }
                ],
                "indexed": "2023-01-31T16:56:50+00:00",
                "status": "current",
                "description": "Metriport-OrgA-1651 Summarization Of Episode Note",
                "mimeType": "application/octet-stream",
                "location": "https://integration.rest.api.commonwellalliance.org/v2/Binary/H4sIAAAAAAAEAJXLMQ6AIAxA0ds4NlRAy2AcnDyAq4s2kQGqCIO3lyuQP778-ZLAi4RQos_fek494ABkFCAgaiINNa0VEJEzvatsK3WJb3l9lvRt0T-FG9ZTjhI45tZxJzcOFgB_QQwPAbUAAAA1",
                "context": {
                    "event": [
                        {
                            "text": "Problem"
                        },
                        {
                            "text": "drug allergy"
                        },
                        {
                            "text": "Nausea"
                        },
                        {
                            "text": "Depressive disorder"
                        },
                        {
                            "text": "Hypothyroidism"
                        },
                        {
                            "text": "Sinusitis"
                        },
                        {
                            "text": "CT of paranasal sinuses"
                        },
                        {
                            "text": "Left maxillary sinus mass"
                        },
                        {
                            "text": "never a smoker"
                        },
                        {
                            "text": "Vital signs"
                        },
                        {
                            "text": "Neoplasm of accessory sinus"
                        },
                        {
                            "text": "Finding of functional performance and activity"
                        },
                        {
                            "text": "Cane"
                        },
                        {
                            "text": "Cognitive function finding"
                        },
                        {
                            "text": "Short term memory deficit"
                        }
                    ],
                    "period": {
                        "start": "2012-10-02T04:00:00Z"
                    }
                }
            }
        }
    ]
}
```

This is the example from [the docs](https://specification.commonwellalliance.org/services/rest-api-reference#8610-documentreference):

```json
{
    "_links": "link relations",
    "resourceType": "DocumentReference",
    "text": {
        "status": "generated",
        "div": "<div>!-- Snipped for Brevity --></div>"
    },
    "contained": [
        {
            "resourceType": "Practitioner",
            "id": "a1",
            "name": {
                "family": [
                    "Dopplemeyer"
                ],
                "given": [
                    "Sherry"
                ]
            },
            "telecom": [
                {
                    "system": "email",
                    "value": "john.doe@healthcare.example.org"
                }
            ],
            "organization": {
                "display": "Cleveland Clinic"
            },
            "role": [
                {
                    "text": "Primary Surgeon"
                }
            ],
            "specialty": [
                {
                    "text": "Orthopedic"
                }
            ]
        },
        {
            "resourceType": "Practitioner",
            "id": "a2",
            "name": {
                "family": [
                    "Smitty"
                ],
                "given": [
                    "Gerald"
                ]
            },
            "telecom": [
                {
                    "system": "email",
                    "value": "john.doe@healthcare.example.org"
                }
            ],
            "organization": {
                "display": "Cleveland Clinic"
            },
            "role": [
                {
                    "text": "Attending"
                }
            ],
            "specialty": [
                {
                    "text": "Orthopedic"
                }
            ]
        }
    ],
    "masterIdentifier": {
        "system": "urn:ietf:rfc:3986",
        "value": "urn:oid:1.3.6.1.4.1.21367.2005.3.7"
    },
    "subject": {
        "reference": "Patient/xcda"
    },
    "type": {
        "coding": [
            {
                "system": "http://loinc.org",
                "code": "34108-1",
                "display": "Outpatient Note"
            }
        ]
    },
    "author": [
        {
            "reference": "#a1"
        },
        {
            "reference": "#a2"
        }
    ],
    "created": "2005-12-24T09:35:00+11:00",
    "indexed": "2005-12-24T09:43:41+11:00",
    "status": "current",
    "description": "Physical",
    "confidentiality": [
        {
            "coding": [
                {
                    "system": "http://ihe.net/xds/connectathon/confidentialityCodes",
                    "code": "1.3.6.1.4.1.21367.2006.7.101",
                    "display": "Clinical-Staff"
                }
            ]
        }
    ],
    "primaryLanguage": "en-US",
    "mimeType": "application/hl7-v3+xml",
    "size": 3654,
    "hash": "2jmj7l5rSw0yVb/vlWAYkK/YBwk=",
    "location": "http://example.org/xds/mhd/Binary/07a6483f-732b-461e-86b6-edb665c45510",
    "context": {
        "event": [
            {
                "coding": [
                    {
                        "system": "http://ihe.net/xds/connectathon/eventCodes",
                        "code": "T-D8200",
                        "display": "Arm"
                    }
                ]
            }
        ],
        "period": {
            "start": "2004-12-23T08:00:00",
            "end": "2004-12-23T08:01:00"
        },
        "facilityType": {
            "coding": [
                {
                    "system": "http://www.ihe.net/xds/connectathon/healthcareFacilityTypeCodes",
                    "code": "Outpatient",
                    "display": "Outpatient"
                }
            ]
        }
    }
}
```

To analyze these I double click on VSCode's open file tab list (on an empty space), it opens a new tab for a new file; select `json` as language; paste it; auto format.

**Review by Orta21 — APPROVED** [review]

File: packages/packages/commonwell-sdk/src/models/document.ts (hunk 4 above)
Comment (Goncharo) [reply packages/packages/commonwell-sdk/src/models/document.ts:49]: Discussed offline, we'll update this as necessary based on what we see in prod.

**Review by Goncharo — APPROVED** [review]
🚢 🚢 🚢

**Pull request merged**